samza-sql/src/main/java/org/apache/samza/sql/impl/ConfigBasedSourceResolverFactory.java

```diff
@@ -44,26 +46,57 @@ public SourceResolver create(Config config) {
   }
 
   private class ConfigBasedSourceResolver implements SourceResolver {
+    private final String SAMZA_SQL_QUERY_TABLE_KEYWORD = "$table";
     private final Config config;
 
     public ConfigBasedSourceResolver(Config config) {
       this.config = config;
     }
 
     @Override
-    public SqlSystemStreamConfig fetchSourceInfo(String source) {
+    public SqlSystemSourceConfig fetchSourceInfo(String source) {
       String[] sourceComponents = source.split("\\.");
+      boolean isTable = false;
+      int systemIdx = 0;
+      int endIdx = sourceComponents.length - 1;
+      int streamIdx = endIdx;
+      boolean invalidQuery = false;
 
-      // This source resolver expects sources of format {systemName}.{streamName}
+      // This source resolver expects sources of format {systemName}.{streamName}[.$table]
       if (sourceComponents.length != 2) {
-        String msg = String.format("Source %s is not of the format {systemName}.{streamName{", source);
+        if (sourceComponents.length != 3 ||
+            !sourceComponents[endIdx].equalsIgnoreCase(SAMZA_SQL_QUERY_TABLE_KEYWORD)) {
+          invalidQuery = true;
+        }
+      } else {
+        if (sourceComponents[0].equalsIgnoreCase(SAMZA_SQL_QUERY_TABLE_KEYWORD) ||
+            sourceComponents[1].equalsIgnoreCase(SAMZA_SQL_QUERY_TABLE_KEYWORD)) {
+          invalidQuery = true;
+        }
+      }
+
+      if (invalidQuery) {
+        String msg = String.format("Source %s is not of the format {systemName}.{streamName}[.%s]", source,
+            SAMZA_SQL_QUERY_TABLE_KEYWORD);
         LOG.error(msg);
         throw new SamzaException(msg);
       }
-      String systemName = sourceComponents[0];
-      String streamName = sourceComponents[1];
 
-      return new SqlSystemStreamConfig(systemName, streamName, fetchSystemConfigs(systemName));
+      if (sourceComponents[endIdx].equalsIgnoreCase(SAMZA_SQL_QUERY_TABLE_KEYWORD)) {
+        isTable = true;
+        streamIdx = endIdx - 1;
+      }
+
+      String systemName = sourceComponents[systemIdx];
+      String streamName = sourceComponents[streamIdx];
```

Comment:
nit: seems the above code can be made simpler/cleaner by using regex and group captures?